okhttp/src/main/java/com/squareup/okhttp/HttpUrl.java

```diff
@@ -0,0 +1,294 @@
+import java.net.URL;
+import java.util.List;
+import java.util.Set;
+
+/**
+ * A <a href="https://url.spec.whatwg.org/">URL</a> with an {@code http} or {@code https} scheme.
+ *
+ * TODO: discussion on canonicalization
+ *
+ * TODO: discussion on encoding-by-parts
+ *
+ * TODO: discussion on this vs. java.net.URL vs. java.net.URI
+ */
+public final class HttpUrl {
+  private HttpUrl(Builder builder) {
+    throw new UnsupportedOperationException();
+  }
+
+  public URL url() {
+    throw new UnsupportedOperationException();
+  }
+
+  public URI uri() throws IOException {
+    throw new UnsupportedOperationException();
+  }
+
+  /** Returns either "http" or "https". */
+  public String scheme() {
+    throw new UnsupportedOperationException();
+  }
+
+  public boolean isHttps() {
+    return scheme().equals("https");
+  }
+
+  public String user() {
+    throw new UnsupportedOperationException();
+  }
+
+  public String encodedUser() {
+    throw new UnsupportedOperationException();
+  }
+
+  /**
+   * Returns the decoded (potentially non-ASCII) hostname. The returned string may contain non-ASCII
+   * characters and is <strong>not suitable</strong> for DNS lookups; for that use {@link
+   * #encodedHost}. For example, this may return {@code ☃.net} which is a user-displayable IDN that
+   * cannot be used for DNS lookups without encoding.
+   */
+  public String host() {
+    throw new UnsupportedOperationException();
+  }
+
+  /**
+   * Returns the host address suitable for use with {@link InetAddress#getAllByName(String)}. May
+   * be a regular host name ({@code android.com}), an IPv4 address ({@code 127.0.0.1}), an IPv6
+   * address ({@code ::1}; note that there are no square braces), or an encoded IDN ({@code
+   * xn--n3h.net}).
+   */
+  public String encodedHost() {
+    throw new UnsupportedOperationException();
+  }
+
+  /**
+   * Returns the explicitly-specified port if one was provided, or the default port for this URL's
+   * scheme. For example, this returns 8443 for {@code https://square.com:8443/} and 443 for {@code
+   * https://square.com/}.
+   */
+  public int port() {
+    throw new UnsupportedOperationException();
+  }
+
+  /**
+   * Returns 80 if {@code scheme.equals("http")}, 443 if {@code scheme.equals("https")} and -1
+   * otherwise.
+   */
+  public static int defaultPort(String scheme) {
+    throw new UnsupportedOperationException();
+  }
+
+  /**
+   * Returns the entire path of this URL, encoded for use in HTTP resource resolution. The
+   * returned path is always nonempty and is prefixed with {@code /}.
+   */
+  public String encodedPath() {
+    throw new UnsupportedOperationException();
+  }
+
+  public List<String> pathSegments() {
+    throw new UnsupportedOperationException();
+  }
+
+  /**
+   * Returns the query of this URL, encoded for use in HTTP resource resolution. The returned string
+   * may be null (for URLs with no query), empty (for URLs with an empty query) or non-empty (all
```

Comment:
Interesting distinction here. You talked about how an explicit port 80 would be dropped for an implicit one. Is there reason for an empty query string to be treating differently than the lack of a query string?